_releases/2022-10-06-1.6.0-released.md

```diff
@@ -0,0 +1,73 @@
+---
+title: Crystal 1.6.0 is released!
+version: 1.6.0
+summary:
+thumbnail: +
+author:
+---
+
+We are delivering a new release with several bugfixes and improvements. Below we list the most important or interesting changes, without mentioning several bugfixes and smaller enhancements. For more details, visit the [changelog](https://github.com/crystal-lang/crystal/releases/tag/1.6.0). Breaking changes are marked with ⚠️.
+
+Pre-built packages are available on [GitHub Releases](https://github.com/crystal-lang/crystal/releases/tag/1.6.0) and our official distribution channels.
+See [crystal-lang.org/install](https://crystal-lang.org/install/) for installation instructions.
+
+## Stats
+
+In this release we included [183 changes since the 1.5.1 release](https://github.com/crystal-lang/crystal/pulls?q=is%3Apr+milestone%3A1.6.0) by 26 contributors. We thank all the effort put into improving the language! ❤️
+
+Below we list the most remarkable changes in the language and stdlib.
+
+## ⚠️ Refactor in console methods
+
+By making some console methods available in Windows, a significant [refactor](https://github.com/crystal-lang/crystal/pull/12352) lead to deprecating three macro methods in `FileDescriptor`: `cooked_from_tc_mode!`, `noecho_from_tc_mode!`, and `raw_from_tc_mode!`. Additionally, methods `#noecho!` and `#raw!` now return `nil`.
+
+## ⚠️ Refactors in `File::Info`
+
+[Another refactor](https://github.com/crystal-lang/crystal/pull/12385) lead to an improvement in the `File` API, which was breaking the abstraction by returning an object intended to be internal. In practice, if for some reason you ended up with an object of type `Crystal::System::FileInfo`, it should now be `File::Info`.
+
+## ⚠️ Improvements in overload ordering
+
+[Union types are now considered first](https://github.com/crystal-lang/crystal/pull/12335) when searching for a suitable overload. Consider the following code:
+
+```crystal
+module Foo(T)
+end
+
+class Bar1
+  include Foo(Int32)
+end
+
+class Bar2
+  include Foo(Int32)
+end
+
+def foo(x : Foo(Int32))
+  'a'
+end
+
+def foo(x : Bar1 | Bar2)
+  true
+end
+
+foo(Bar1.new)
+```
+
+Before 1.6.0 it would print `'a'`, meaning the generic case was considered first. Now it correctly prints `true`.
```

Comment:
We should clearly point out that this bug fix involves a hard breaking change that can lead to programs not building any more. And give advise on what to do when you encounter such a problem (probably restructure the overload type restrictions?). I suppose we didn't put a flag to disable this? @HertzDevil

Reply:
A migration guide would be something like:

* This happens whenever a parameter has a `Union` restriction in one overload, and the same parameter in another overload has a restriction that isn't a `Path` or an `Underscore`.
* To preserve the old ordering, remove the union overload;
* To ensure the new ordering applies to older releases, move the union overload above the generic one. (This is because previous Crystal versions consider the two overloads unordered.)

Reply:
What do you mean by "remove the union overload"? Remove the type restriction with the union?

Reply:
No, just straight up deleting the `foo(x : Bar1 | Bar2)` overload because it never matches in existing versions.